### ISSUE TITLE
Chapter 4: mini-batch gradient descent

### ISSUE THREAD

**Issue by vd1** [ageron/handson-ml#179]
In the 04_notebook, when defining the mini-batch gradient, would it be better perhaps to divide by minibatch_size?

here is the defining line:

```gradients = 2 * xi.T.dot(xi.dot(theta) - yi)```

This way the behaviour would be the same
as 1) full gradient descent (GD) when the mini-batch size is the full data size 
and 2) SGD when the mini-batch size is 1

**Comment by ageron** [comment]
Hi @vd1 , thanks for your interesting suggestion.  To compute the gradients, we need to divide by the size of the batch, not the size of the full training set.  In the case of stochastic gradient descent, the batch size is equal to 1, so that's why we don't see it here.  But if we were doing mini-batch or batch gradient descent, then we would definitely have to divide by the size of the batch.
If you divided by the full training set size every time, then you would need to change the learning rate whenever you change the batch size.
Wdyt?

**Comment by vd1** [comment]
Agreed! I think we are saying the same thing. But your code for the mini-batch case (cell 23, line 22) does not divide by the mini-batch size in the current version.

**Comment by ageron** [comment]
Oops, I wasn't looking in the right place: I searched for `gradients = 2 * xi.T.dot(xi.dot(theta) - yi)` in the notebook and I found it in the Stochastic Gradient Descent section; I didn't realize it was also present in the Mini-batch Gradient Descent part.  Sorry about that!
Fixing this now! Thanks a lot for all your help on this project, I appreciate it greatly.

**Comment by ageron** [comment]
Note that in order to keep the same result as before, I multiplied the learning rate by the batch size (20).
